### ISSUE TITLE
[Lab3] 3.3-3.4 Loss not Decreasing

### ISSUE THREAD

**Issue by chrisdim** [slp-ntua/slp-labs#61]
Καλησπέρα,
Κατά την εκπαίδευση του νευρωνικού παρατηρώ ότι ανά εποχή το loss που αναγράφεται δεν φαίνεται να μειώνεται αντιθέτως κυμαίνεται σε περίπου σταθερές τιμές. Και παρακάτω στο evaluation τόσο το loss του train όσο και του test είναι περίπου σταθερά για κάθε εποχή, με το πρώτο να μειώνεται σχεδόν καθόλου από 48 τελικώς σε 40 και το δεύτερο να παραμένει συνεχώς ίσο με περίπου 3.5. Κάποιες τιμές μάλιστα του loss παίρνουν την τιμή inf (???).
Χρησιμοποιώ το model glove.twitter.27B.100d.txt πάνω στο dataset MR, με BCEWithLogitsLoss loss function (στην οποια εχω φροντισει να περναω τα παντα ως logits) και adam optimizer με learning_rate = 0.001. Επίσης batch_size = 128 και epochs = 50. Υπάρχει κάποιο σφάλμα που μπορεί να έχω κάνει;
Ευχαριστώ.

**Comment by pxaris** [comment]
Το `BCEWithLogitsLoss` περιμένει τα "targets" να είναι αριθμοί μεταξύ 0 και 1 που έχουν το ίδιο shape με το "input" (βλ. [BCEWithLogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html).)
Στην περίπτωσή σου, βεβαιώσου ότι μετασχηματίζεις τo καθε label σε one-hot vector πριν εφαρμόσεις το loss αυτό.

**Comment by chrisdim** [comment]
Καλημέρα είχα ήδη φροντίσει για τη σωστή μορφή των logits των labels δηλαδή την κλάση 0 την έχω [1.0, 0.0] και την κλαση 1 [0.0, 1.0]. Τα logits που μου επιστρέφει το μοντέλο ως εκτίμηση όμως δεν φαίνεται να περιορίζονται στο διάστημα [0,1]. πχ βγάζει τιμές tensor([-1.5809,  1.5230]), tensor([ 0.8227, -0.7958]). Υπάρχει πρόβλημα σε αυτήν τη μορφή;

**Comment by pxaris** [comment]
Όχι, δεν είναι πρόβλημα αυτό. Πειραματίσου με τις υπερπαραμέτρους του μοντέλου, δες αν μεταβάλλεται η κατάσταση, αλλιώς πάρε κομμάτι κομμάτι τον κώδικα σου για να εντοπίσεις σε ποιό σημείο υπάρχει κάποιο bug.

**Comment by chrisdim** [comment]
Καλησπέρα,
Έχοντας επιλέξει κατάλληλη τιμή για το batch size κατάφερα για 100 εποχές να μειώσω το training loss μέχρι την τιμή 0.34 επιτυγχάνοντας accuracy 83%. Ωστόσο, το test loss φαίνεται να αυξάνεται καθώς γίνεται το training, παρόλο που επιτυγχάνεται accuracy 70%. Είναι μια τέτοια συμπεριφορά αναμενόμενη;

**Comment by pxaris** [comment]
Είναι αναμενόμενο αλλά όχι επιθυμητό. Το μοντέλο εχει κάνει **overfitting** στα δεδομένα εκπαίδευσης, πράγμα που σημαίνει ότι έχει μειωθεί η γενικευτική του ικανότητα (άυξηση του generalization error που μετριέται μέσω του test loss). 
Η λύση είναι η εφαρμογή τεχνικών **regularization**, μια από τις οποίες είναι το "early-stopping" το οποίο θα υλοποιηθεί (απλοϊκά) στην άσκηση μέσω της αποθήκευσης του "ckeckpoint" του μοντέλου που έχει το χαμηλότερο test loss κατά την εκπαίδευση (βλ. Ερ-5.1 Εργαστηριακής).
